### PR TITLE
feat(inference): multi-GPU auto-layout and cross-GPU routing (PR2)

### PR DESCRIPTION
**Pull request by greenstephen**
## Summary
- Add VRAM-aware `gpu_mapping=auto` layout for multi-GPU CUDA systems (DiT stack on one GPU, LM on another).
- Route conditioning tensors to the DiT GPU at inference phase boundaries (preprocess → diffusion → VAE decode).
- Initialize the LM on `device_map.lm` with per-GPU vLLM memory budgeting.

## Stacked on
- **Depends on #1262** (PR1: ComponentDeviceMap). Merge PR1 first; after that this PR will show only the PR2 commit (`a38527f`).

## Scope
- **In:** `device_map.py` auto-layout, cross-GPU tensor routing, LM device wiring, tests
- **Out:** CLI `--gpu-mapping` flag, Gradio UI panel, LM tensor parallelism (PR3–PR4)

## Risk and Compatibility
- **Target path:** CUDA multi-GPU inference with `ACESTEP_GPU_MAPPING=auto` or explicit multi-component maps
- **Non-target paths unchanged:** Single-GPU behavior when mapping is unset; MPS/XPU/CPU unchanged

## Regression Checks
- [x] `acestep.test_device_map` — 18 tests pass
- [x] `acestep.core.generation.handler.init_service_test` — 83/84 pass (1 pre-existing failure)

## Test plan
- [ ] `ACESTEP_GPU_MAPPING=auto` on 2×24GB — XL DiT + 4B LM init without OOM
- [ ] `ACESTEP_GPU_MAPPING=dit:0,vae:0,text_encoder:0,lm:1` — generate a short track
- [ ] Single GPU, no mapping — no behavior change vs main+PR1

Made with [Cursor](https://cursor.com)

## Summary by CodeRabbit

* **New Features**
  * Added multi-component, multi-GPU device mapping with automatic VRAM-aware placement, plus a public device-mapping API for GPU discovery, layout planning, and runtime status.
  * Extended LLM startup and UI initialization to respect the mapped LM device and optional GPU mapping environment setting.
* **Bug Fixes**
  * Improved device placement for conditioning inputs, intermediate tensors, silence handling, and generation timesteps to target the correct component device.
  * Fixed CUDA handling to consistently respect explicit CUDA indices during memory budgeting and vLLM/FlashAttention setup.
* **Tests**
  * Expanded coverage for device-map parsing/resolution, DiT payload routing, and CUDA index–sensitive initialization behavior.

### CODE REVIEW

File: acestep/core/generation/handler/init_service_memory_basic.py [hunk 1]
```diff
@@ -152,10 +152,15 @@ def _has_quantized_params(self, module):
...
         if hasattr(self, "silence_latent") and self.silence_latent is not None:
-            if not self._is_on_target_device(self.silence_latent, self.device):
-                self.silence_latent = self.silence_latent.to(self.device).to(self.dtype)
+            target = (
+                self._get_component_device("dit")
+                if getattr(self, "device_map", None) is not None
+                else self.device
+            )
+            if not self._is_on_target_device(self.silence_latent, target):
+                self.silence_latent = self.silence_latent.to(target).to(self.dtype)
```
[the remark this answers is not in the document]
Comment (greenstephen) [reply acestep/core/generation/handler/init_service_memory_basic.py:176]: Fixed in b6d4061 / 3a80d6a. `_ensure_silence_latent_on_device()` now uses
`_tensor_on_exact_device()` so we compare the full CUDA device (e.g. cuda:3),
not just backend type via `_is_on_target_device()`. Added unit tests.

File: acestep/device_map/layout.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        if lm_gpu is None:
+            return LayoutError(
+                f"No GPU has {lm_need_gb:.1f}GB free for LM ({request.lm_model_path or 'default'})",
+                suggestions=(
+                    "Use a smaller LM model",
+                    "Pass gpu_mapping with lm on a specific GPU",
+                    "Disable LM initialization",
+                ),
+            )
+        lm_device = f"cuda:{lm_gpu.logical_index}"
```
[the remark this answers is not in the document]
Comment (greenstephen) [reply acestep/device_map/layout.py:107]: Already addressed in 9d998c6: `_lm_free_vram_gb()` subtracts `dit_need_gb`
when LM falls back to the DiT GPU, so the layout check uses remaining headroom
instead of raw `free_vram_gb`.

File: acestep/models/common/apg_guidance.py [hunk 3]
```diff
@@ -19,15 +19,19 @@ def project(
...
 
     v0, v1 = v0.double(), v1.double()
     v1 = torch.nn.functional.normalize(v1, dim=dims)
     v0_parallel = (v0 * v1).sum(dim=dims, keepdim=True) * v1
     v0_orthogonal = v0 - v0_parallel
-    return v0_parallel.to(dtype).to(device_type), v0_orthogonal.to(dtype).to(device_type)
+    return (
+        v0_parallel.to(device=device, dtype=dtype),
+        v0_orthogonal.to(device=device, dtype=dtype),
+    )
```
[the remark this answers is not in the document]
Comment (greenstephen) [reply acestep/models/common/apg_guidance.py:33]: Already addressed in b0ead26: we capture `target_device` before the MPS CPU
fallback and return with `.to(device=target_device, dtype=dtype)` so tensors
go back to MPS, not CPU.

File: acestep/device_map.py [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+    if parsed is None:
+        return ComponentDeviceMap.from_single_device(normalized_device)
+    return parsed
+
+
+def log_device_map(device_map: ComponentDeviceMap) -> None:
+    """Emit a startup log line describing the active component layout."""
+    logger.info("[device_map] Active layout: {}", device_map.summary())
+    if device_map.is_multi_device():
+        logger.info("[device_map] Multi-device placement enabled with cross-GPU routing")
```
[the remark this answers is not in the document]
Comment (greenstephen) [reply acestep/device_map.py]: Addressed in 500e04a: split device_map.py into acestep/device_map/{types,parsing,layout,discovery,resolve,status,devices}.py with a facade __init__.py. All submodules are ≤104 LOC. Public imports unchanged.

File: acestep/device_map.py [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+                ),
+            )
+        lm_device = f"cuda:{lm_gpu.logical_index}"
+
+    return ComponentDeviceMap(
+        dit=dit_device,
+        vae=dit_device,
+        text_encoder=dit_device,
+        lm=lm_device,
+    )
```
[the remark this answers is not in the document]
Comment (greenstephen) [reply acestep/device_map.py]: Duplicate of the resolved layout.py thread — _lm_free_vram_gb() in layout.py
reserves dit_need_gb before LM fallback to the DiT GPU. File device_map.py no
longer exists after the package split.

**Comment by pokepress** [comment]
So, just to put some comments specific to this PR here, the code in llm_inference.py only respects the LM device setting if the device setting is "cuda" as opposed to "auto". Additionally, it seems like the backend needs to be set to PT since VLLM doesn't seem to be respecting the setting and uses the same GPU for all models. Let me know if you have other things you want me to try.

**Comment by greenstephen** [comment]
Thanks — good catches, and useful to keep them on this PR.

auto vs cuda: Agreed. In llm_inference.initialize, device == "auto" is collapsed to bare "cuda" (no index), so a mapped cuda:1 never sticks on that path. Related: several later checks use device == "cuda" / device != "cuda", which also break indexed devices. One concrete footgun is the vLLM gate:

if backend == "vllm" and device != "cuda":
    # falls back to pt
So "cuda:1" is treated as “not CUDA for vLLM” and gets forced to PT, while bare "cuda" stays on vLLM and lands on the default GPU (usually 0).

vLLM same-GPU behavior: That matches what we’re seeing. Until indexed CUDA is handled cleanly in the vLLM init path, --backend pt is the reliable way to keep the LM on lm:N. We’ll treat proper vLLM + cuda:N as part of the same fix.

Things worth trying:

--backend pt with --gpu-mapping "dit:0,vae:0,text_encoder:0,lm:1" and --init-service true
Confirm logs show Loading LLM to cuda:1 (not bare cuda)
Watch nvidia-smi during LM vs DiT phases
We’ll fix the Gradio/device_map wiring plus the bare-"cuda" comparisons in a follow-up on this stack. Appreciate the testing.

**Comment by greenstephen** [comment]
@pokepress Thanks again — you were right about the LM landing on bare `cuda`.

I’ve pushed a fix on this branch (`ec02094`) that addresses:

1. **Gradio init ordering** — `lm_device` is now taken from `device_map` **after** DiT `initialize_service`, so `lm:1` isn’t lost on first UI init when the device dropdown is `auto`.
2. **`llm_inference` CUDA indexing** — `auto` / bare `cuda` normalize to `cuda:0`; mapped `cuda:N` is preserved. The vLLM gate now uses `is_cuda_device()` so `cuda:1` is no longer treated as “not CUDA” and forced to PT.
3. **Related** — text-encoder token ids are moved onto the text-encoder device before encode (helps split maps).

**Please retry** (after pulling this branch):

```bash
uv run acestep \
  --init-service true \
  --config-path acestep-v15-xl-sft \
  --lm_model_path acestep-5Hz-lm-4B \
  --init-llm true \
  --gpu-mapping "dit:0,vae:0,text_encoder:0,lm:1"
